src/RecursiveIteratorAggregateIterator.php

```diff
@@ -0,0 +1,80 @@
+<?php
+
+declare(strict_types=1);
+
+namespace loophp\iterators;
+
+use Generator;
+use Iterator;
+use IteratorAggregate;
+use RuntimeException;
+use Traversable;
+
+use function count;
+
+/**
+ * @template TKey
+ * @template T
+ *
+ * @implements IteratorAggregate<int, T>
+ */
+class RecursiveIteratorAggregateIterator implements IteratorAggregate
+{
+    /**
+     * @var list<Iterator>
+     */
+    private array $stack = [];
+
+    /**
+     * @param Traversable<TKey,T> $input
+     */
+    public function __construct(private readonly Traversable $input) {}
+
+    public function getDepth(): int
+    {
+        return count($this->stack);
+    }
+
+    /**
+     * @return Generator<int,T>
+     */
+    public function getIterator(): Generator
+    {
+        $iterator = $this->findIterator($this->input);
+        $this->stack = [];
```

Comment:
Let's keep the object immutable, this should not happen

Reply:
We can`t have state tracking (getDepth) on immutable iterator, we can add exception here